### PR TITLE
common: inf and NaN are invalid in JSON

### PR DESCRIPTION
**Pull request by sajibreadd**
Fixes: https://tracker.ceph.com/issues/66215

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/common/Formatter.cc [hunk 1]
```diff
@@ -296,6 +296,17 @@ void JSONFormatter::finish_pending_string()
   }
 }
 
+void JSONFormatter::add_value(std::string_view name, double val) {
```
Comment (sebastian-philipp) [inline src/common/Formatter.cc:299]: wat to add a test to https://github.com/ceph/ceph/blob/main/src/test/common/test_json_formatter.cc ?

**Comment by sebastian-philipp** [comment]
@batrick do you want to review this?

File: src/common/Formatter.cc (hunk 1 above)
Comment (sajibreadd) [reply src/common/Formatter.cc:299]: Added the test.

File: src/test/common/test_json_formatter.cc [hunk 2]
```diff
@@ -79,3 +79,20 @@ TEST(formatter, utime)
   EXPECT_EQ(input.sec(), output.sec());
   EXPECT_EQ(input.nsec(), output.nsec());
 }
+
+TEST(formatter, bug_66215)
```
Comment (batrick) [inline src/test/common/test_json_formatter.cc:83]:
```suggestion
TEST(formatter, dump_inf_or_nan)
```

File: src/common/Formatter.cc (hunk 1 above)
Comment (batrick) [inline src/common/Formatter.cc:299]:
```suggestion
template <typename T, typename = std::enable_if_t<std::is_floating_point_v<T>>>
void JSONFormatter::add_value(std::string_view name, T val) {
```

File: src/common/Formatter.h [hunk 3]
```diff
@@ -232,7 +232,7 @@ namespace ceph {
     void print_quoted_string(std::string_view s);
     void print_name(std::string_view name);
     void print_comma(json_formatter_stack_entry_d& entry);
-
+    void add_value(std::string_view name, double val);
```
Comment (batrick) [inline src/common/Formatter.h:235]:
```suggestion
template <typename T, typename = std::enable_if_t<std::is_floating_point_v<T>>>
    void add_value(std::string_view name, T val);
```

File: src/common/Formatter.cc [hunk 4]
```diff
@@ -296,6 +296,17 @@ void JSONFormatter::finish_pending_string()
...
+  CachedStackStringStream css;
+  if (std::isinf(val) || std::isnan(val)) {
+    *css << "null";
```
Comment (batrick) [inline src/common/Formatter.cc:302]: I would prefer we use a string like "[+-]inf" and "NaN". Then it can be decoded if desired to the original representation.

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/common/Formatter.cc (hunk 4 above)
Comment (sajibreadd) [reply src/common/Formatter.cc:302]: Are you looking for quoted string "[+-]inf"?

File: src/common/Formatter.h (hunk 3 above)
Comment (sajibreadd) [reply src/common/Formatter.h:235]: Shouldn't be it ambiguous if keeping both functions?
```
    template <typename T, typename = std::enable_if_t<std::is_floating_point_v<T>>>
    void add_value(std::string_view name, T val);

    template <typename T>
    void add_value(std::string_view name, T val);
```

File: src/common/Formatter.cc (hunk 4 above)
Comment (sebastian-philipp) [reply src/common/Formatter.cc:302]: the JSON RFC doesn't know things like `-inf` or similar. We end up with client libraries that just bail out if they encounter invalid JSON.

Comment (sajibreadd) [reply src/common/Formatter.cc:302]: What about quoted string?

Comment (patrakov) [reply src/common/Formatter.cc]: I would vote for placing such exceptional values into a differently named field, which is, however, impossible to do in this function. In any case, making a field sometimes (especially in rare cases) have a different data type (a string instead of a number) is bad, as it would cause crashes in type-aware clients that implicitly apply a schema.

In short, this function is not a proper place to make a policy on encoding non-numbers in places where numbers are expected.

File: src/common/Formatter.h (hunk 3 above)
Comment (batrick) [reply src/common/Formatter.h:235]: Ya, nevermind.

File: src/common/Formatter.cc (hunk 4 above)
Comment (batrick) [reply src/common/Formatter.cc]: Is `null` any better?

**Review by batrick — CHANGES_REQUESTED** [review]
Commit title `JSON Parser does not allow inf and NaN. Converted it to null.` to `common/Formatter: dump inf/nan as strings`

File: src/common/Formatter.cc (hunk 4 above)
Comment (patrakov) [reply src/common/Formatter.cc]: On paper, yes: with Rust and serde_json, one can directly serialize a "floating point or `null`" into `Option<f32>`.

In another possible opinion, any NaNs or infinities here should be treated as errors (i.e., `ceph_assert`) and all callers fixed up. Of course, this is a monumental task, that's why I am not suggesting this for real, but writing `null` and logging a backtrace might be something to consider.

Comment (batrick) [reply src/common/Formatter.cc]: Very well, let's stick with `null`.

**Comment by sajibreadd** [comment]
Is this PR going for the test? Also I think we need to backport this PR.

File: src/test/common/test_json_formatter.cc [hunk 5]
```diff
@@ -79,3 +79,22 @@ TEST(formatter, utime)
...
+  formatter.dump_float("nan_val", std::numeric_limits<double>::quiet_NaN());
+  formatter.dump_float("nan_val_alt", std::nan(""));
+  formatter.dump_float("double_max", std::numeric_limits<double>::max());
+  formatter.dump_float("double_min", std::numeric_limits<double>::min());
+  formatter.close_section();
+  bufferlist bl;
+  formatter.flush(bl);
+  JSONParser parser;
+  cout << std::string(bl.c_str(), bl.length()) << endl;
+  EXPECT_TRUE(parser.parse(bl.c_str(), bl.length()));
```
Comment (batrick) [inline src/test/common/test_json_formatter.cc:99]: should we not also verify that the values are `null`?

**Review by batrick — CHANGES_REQUESTED** [review]

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
jenkins test make check arm64

File: src/common/Formatter.cc [hunk 6]
```diff
@@ -296,7 +296,18 @@ void JSONFormatter::finish_pending_string()
   }
 }
 
-template <class T>
+void JSONFormatter::add_value(std::string_view name, double val) {
+  CachedStackStringStream css;
+  if (std::isinf(val) || std::isnan(val)) {
```
Comment (patrakov) [inline src/common/Formatter.cc:301]: Would the shorter `!std::isfinite(val)` form work here?

**Comment by sajibreadd** [comment]
> jenkins test make check arm64
@batrick sorry confused a bit. This field is not required. Do I need to do any adjustment to get check mark here?

File: src/common/Formatter.cc (hunk 6 above)
Comment (sajibreadd) [reply src/common/Formatter.cc:301]: Sorry, missed your comment somehow.
It does work but do you have any special observation for using that instead of `std::isinf(val)` ?

Comment (patrakov) [reply src/common/Formatter.cc:301]: It's shorter, expresses the intent more directly, enumerates goodness not badness,... i.e. purely aesthetical reasons.

**Comment by sajibreadd** [comment]
jenkins test make check

**Comment by sajibreadd** [comment]
jenkins test docs

**Comment by sajibreadd** [comment]
jenkins test api
jenkins test windows

**Comment by sajibreadd** [comment]
jenkins test api

**Comment by sajibreadd** [comment]
jenkins test windows

**Comment by sajibreadd** [comment]
jenkins test make check arm64

**Comment by sajibreadd** [comment]
jenkins test api

File: src/common/Formatter.h [hunk 7]
```diff
@@ -232,8 +232,9 @@ namespace ceph {
     void print_quoted_string(std::string_view s);
     void print_name(std::string_view name);
     void print_comma(json_formatter_stack_entry_d& entry);
+    void add_value(std::string_view name, double val);
 
-    template <class T>
+    template <typename T>
```
Comment (batrick) [inline src/common/Formatter.h:237]: undo please

File: src/common/Formatter.cc [hunk 8]
```diff
@@ -296,7 +296,18 @@ void JSONFormatter::finish_pending_string()
   }
 }
 
-template <class T>
```
Comment (batrick) [inline src/common/Formatter.cc:299]: undo this change

File: src/test/common/test_json_formatter.cc [hunk 9]
```diff
@@ -79,3 +79,27 @@ TEST(formatter, utime)
...
+  formatter.flush(bl);
+  JSONParser parser;
+  parser.parse(bl.c_str(), bl.length());
+  cout << std::string(bl.c_str(), bl.length()) << endl;
+  EXPECT_TRUE(parser.parse(bl.c_str(), bl.length()));
+  EXPECT_EQ(parser.find_obj("positive_infinity")->get_data(), "null");
+  EXPECT_EQ(parser.find_obj("negative_infinity")->get_data(), "null");
+  EXPECT_EQ(parser.find_obj("nan_val")->get_data(), "null");
+  EXPECT_EQ(parser.find_obj("nan_val_alt")->get_data(), "null");
+}
```
Comment (batrick) [inline src/test/common/test_json_formatter.cc:105]: missing terminal newline

**Review by batrick — CHANGES_REQUESTED** [review]
minor nits, otherwise lgtm

**Comment by sajibreadd** [comment]
jenkins test make check

**Comment by sajibreadd** [comment]
jenkins test make check

**Comment by sajibreadd** [comment]
jenkins test api

**Comment by sajibreadd** [comment]
jenkins test api

**Comment by sajibreadd** [comment]
@batrick Is there anything I need to do in this PR?

File: src/test/common/test_json_formatter.cc (hunk 9 above)
Comment (ifed01) [inline src/test/common/test_json_formatter.cc:105]: please verify double_max/double_min values as well

**Comment by batrick** [comment]
@sajibreadd did you mean to close this? Please reopen?

**Comment by sajibreadd** [comment]
> @sajibreadd did you mean to close this? Please reopen?

Here is some mistake I did. I merged main into the branch by mistake(by syncing the fork by mistake). As there is a lot of commit to drop that's why here is the things I have done,
1. created a new branch.
2. cherrypicked the commit.
3. pushed the branch upstream
4. deleted the old branch named wip-66215
5. renamed the newly created branch wip-66215

Now github is not allowing me to reopen the PR. Apologies for the mistakes. Now what should be the best things to do? create another PR and have reference of this PR?
